component/pyroscope/ebpf/ebpfspy/sd/target.go

```diff
@@ -0,0 +1,220 @@
+	if k8sNamespace != "" && k8sContainer != "" {
+		return fmt.Sprintf("ebpf/%s/%s", k8sNamespace, k8sContainer)
+	}
+	dockerContainer := target["__meta_docker_container_name"]
+	if dockerContainer != "" {
+		return dockerContainer
+	}
+	return "unspecified"
+}
+
+func (t *Target) Labels() (uint64, labels.Labels) {
+	if !t.fingerprintCalculated {
+		t.fingerprint = t.labels.Hash()
+		t.fingerprintCalculated = true
+	}
+	return t.fingerprint, t.labels
+}
+
+type containerID string
+
+type TargetFinder interface {
+	FindTarget(pid uint32) *Target
+	DebugInfo() []string
+	Update(args TargetsOptions)
+}
+type TargetsOptions struct {
+	Targets            []discovery.Target
+	TargetsOnly        bool
+	DefaultTarget      discovery.Target
+	ContainerCacheSize int
+}
+
+type targetFinder struct {
+	l          log.Logger
+	cid2target map[containerID]*Target
+
+	// todo make it never evict during a reset
+	containerIDCache *lru.Cache[uint32, containerID]
+	defaultTarget    *Target
+	fs               fs.FS
+}
+
+func (tf *targetFinder) Update(args TargetsOptions) {
+	tf.setTargets(args)
+	tf.resizeContainerIDCache(args.ContainerCacheSize)
+}
+
+func NewTargetFinder(fs fs.FS, l log.Logger, options TargetsOptions) (TargetFinder, error) {
+	containerIDCache, err := lru.New[uint32, containerID](options.ContainerCacheSize)
+	if err != nil {
+		return nil, fmt.Errorf("containerIDCache create: %w", err)
+	}
+	res := &targetFinder{
+		l:                l,
+		containerIDCache: containerIDCache,
+		fs:               fs,
+	}
+	res.setTargets(options)
+	return res, nil
+}
+
+func (tf *targetFinder) setTargets(opts TargetsOptions) {
+	_ = level.Debug(tf.l).Log("msg", "set targets", "count", len(opts.Targets))
+	containerID2Target := make(map[containerID]*Target)
+	for _, target := range opts.Targets {
+		cid := containerIDFromTarget(target)
+		if cid != "" {
+			t, err := NewTarget(cid, target)
+			if err != nil {
+				_ = level.Error(tf.l).Log(
+					"msg", "target skipped",
+					"target", target.Labels().String(),
+					"err", err,
+				)
+				continue
+			}
+			containerID2Target[cid] = t
+		}
+	}
+	if len(opts.Targets) > 0 && len(containerID2Target) == 0 {
+		_ = level.Warn(tf.l).Log("msg", "No container IDs found in targets")
+	}
+	tf.cid2target = containerID2Target
+	if opts.TargetsOnly {
+		tf.defaultTarget = nil
+	} else {
+		t, err := NewTarget("", opts.DefaultTarget)
+		if err != nil {
+			_ = level.Error(tf.l).Log(
+				"msg", "default target skipped",
+				"target", opts.DefaultTarget.Labels().String(),
+				"err", err,
+			)
+			tf.defaultTarget = nil
+		} else {
+			tf.defaultTarget = t
+		}
+	}
+	_ = level.Debug(tf.l).Log("msg", "created targets", "count", len(tf.cid2target))
+}
+
+func (tf *targetFinder) FindTarget(pid uint32) *Target {
+	res := tf.findTarget(pid)
+	if res != nil {
+		return res
+	}
+	return tf.defaultTarget
+}
+
+func (tf *targetFinder) findTarget(pid uint32) *Target {
+	cid, ok := tf.containerIDCache.Get(pid)
+	if ok && cid != "" {
```

Comment:
What are the chances that a PID that has no associated containerID gets one at some point ? I wonder if we should cache miss at least for a few hours ?